### PR TITLE
Allow external bindgen for fips; more prefixing; bump aws-lc-fips-sys to v0.12.8

### PR DESCRIPTION
**Pull request by justsmth**
### Issues
Mostly Resolves: #366 

### Description of changes:
* Bumps crate version to v0.12.8. Bindings generated in [this run](https://github.com/aws/aws-lc-rs/actions/runs/8638442246)
* Adds prefixing to symbols from rust_wrapper.c.
* When no pre-generated bindings are available, build logic looks for and uses external `bindgen` executable for bindings generation.
* Adds CI to test use of external bindgen.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: aws-lc-fips-sys/CMakeLists.txt [hunk 1]
```diff
@@ -60,4 +60,8 @@ if (BORINGSSL_PREFIX)
         set(TARGET_PREFIX "lib${BORINGSSL_PREFIX}")
     endif()
     set_my_target_properties(PREFIX ${TARGET_PREFIX})
+
+    # This BORINGSSL_PREFIX has an "_" appended, so we must remove it
+    string(REGEX REPLACE "_$" "" BORINGSSL_PREFIX_MACRO ${BORINGSSL_PREFIX})
+    add_definitions(-DBORINGSSL_PREFIX=${BORINGSSL_PREFIX_MACRO})
```
Comment (skmcgrail) [inline aws-lc-fips-sys/CMakeLists.txt:66]: `set(BORINGSSL_PREFIX_MACRO "${BORINGSSL_PREFIX}")` and then using `BORINGSSL_PREFIX_MACRO` for the in/out to the regex replace might be clearer.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -219,6 +219,37 @@ jobs:
         # See: https://github.com/rust-lang/cargo/issues/8531
         run: AWS_LC_FIPS_SYS_STATIC=${{ matrix.static }} cargo test --tests --features fips
 
+  build-env-fips-external-bindgen-test:
+    if: github.repository == 'aws/aws-lc-rs'
```
Comment (skmcgrail) [inline .github/workflows/tests.yml:223]: NP; `repository_owner`.

Comment (WillChilds-Klein) [reply .github/workflows/tests.yml:223]: ^ `githug.repository == 'aws'`?

File: aws-lc-fips-sys/builder/main.rs [hunk 3]
```diff
@@ -423,3 +455,131 @@ fn setup_include_paths(out_dir: &Path, manifest_dir: &Path) -> PathBuf {
...
+        header.as_str(),
+        "--",
+    ]);
+    clang_args
+        .iter()
+        .for_each(|x| bindgen_params.push(x.as_str()));
+    let cmd_params: Vec<OsString> = bindgen_params.iter().map(OsString::from).collect();
+    let cmd_params: Vec<&OsStr> = cmd_params.iter().map(OsString::as_os_str).collect();
+
+    let result = test_command("bindgen".as_ref(), cmd_params.as_ref());
```
Comment (skmcgrail) [inline aws-lc-fips-sys/builder/main.rs:528]: `test_command` -> `execute_command`

File: aws-lc-fips-sys/builder/main.rs [hunk 4]
```diff
@@ -423,3 +455,131 @@ fn setup_include_paths(out_dir: &Path, manifest_dir: &Path) -> PathBuf {
...
+        sym_prefix = if target_os().to_lowercase() == "macos" || target_os().to_lowercase() == "ios"
+        {
+            format!("_{prefix_str}_")
+        } else {
+            format!("{prefix_str}_")
+        };
+        bindgen_params.extend(vec!["--prefix-link-name", sym_prefix.as_str()]);
+    }
+
+    bindgen_params.extend(vec![
```
Comment (skmcgrail) [inline aws-lc-fips-sys/builder/main.rs:500]: Add comment about keeping these flag arguments in-sync with the library version.

**Review by skmcgrail — DISMISSED** [review]

File: aws-lc-fips-sys/include/rust_wrapper.h [hunk 5]
```diff
@@ -6,26 +6,29 @@
...
-#define BORINGSSL_SHARED_LIBRARY
-#define AWS_LC_FIPS_SYS_EXPORT __declspec(dllexport)
-#else
-#define AWS_LC_FIPS_SYS_EXPORT __attribute__((visibility("default")))
-#endif
-
 #include <openssl/err.h>
 
+#define AWS_LC_WRAPPER_ADD_PREFIX(a, b) AWS_LC_WRAPPER_ADD_PREFIX_INNER(a, b)
+#define AWS_LC_WRAPPER_ADD_PREFIX_INNER(a, b) a ## _ ## b
```
Comment (samuel40791765) [inline aws-lc-fips-sys/include/rust_wrapper.h:12]: Is there a particular reason why we can't just use
```
#define AWS_LC_WRAPPER_ADD_PREFIX(a, b) a ## _ ## b
```

File: aws-lc-fips-sys/include/rust_wrapper.h [hunk 6]
```diff
@@ -6,26 +6,29 @@
...
-
 // The following functions are wrappers over inline functions and macros in
 // BoringSSL, which bindgen cannot currently correctly bind. These wrappers
 // ensure changes to the functions remain in lockstep with the Rust versions.
-AWS_LC_FIPS_SYS_EXPORT int ERR_GET_LIB_RUST(uint32_t packed_error);
-AWS_LC_FIPS_SYS_EXPORT int ERR_GET_REASON_RUST(uint32_t packed_error);
-AWS_LC_FIPS_SYS_EXPORT int ERR_GET_FUNC_RUST(uint32_t packed_error);
+OPENSSL_EXPORT int ERR_GET_LIB_RUST(uint32_t packed_error);
+OPENSSL_EXPORT int ERR_GET_REASON_RUST(uint32_t packed_error);
+OPENSSL_EXPORT int ERR_GET_FUNC_RUST(uint32_t packed_error);
```
Comment (samuel40791765) [inline aws-lc-fips-sys/include/rust_wrapper.h:31]: Curious if `BORINGSSL_IMPLEMENTATION` and `BORINGSSL_SHARED_LIBRARY` needed to be defined for this to work in the end?
Just following up from the thread about this a while back: https://github.com/aws/aws-lc-rs/pull/309/files#r1465488911

**Review by samuel40791765 — APPROVED** [review]

File: aws-lc-fips-sys/include/rust_wrapper.h (hunk 5 above)
Comment (justsmth) [reply aws-lc-fips-sys/include/rust_wrapper.h]: The same is done for `BORINGSSL_ADD_PREFIX` -- it allows for [macro expansion on the arguments](https://github.com/aws/aws-lc/blob/4cd6d216e6f13c1a3528dbbf68c2cc46f593ef44/util/make_prefix_headers.go#L109). But, in our limited case, this indirection is likely not necessary.

**Review by samuel40791765 — APPROVED** [review]

**Review by skmcgrail — APPROVED** [review]

**Pull request merged**